anvill/src/Lifters/DataLifter.cpp

```diff
@@ -73,82 +73,110 @@ DataLifter::GetOrDeclareData(const GlobalVarDecl &decl,
+  const auto &dl = options.module->getDataLayout();
+  const auto type = remill::RecontextualizeType(decl.type, context);
+
+  std::vector<uint8_t> bytes;
+  llvm::Constant *value = nullptr;
+  bool bytes_accessable = false;
+
   std::stringstream ss2;
   ss2 << "var_" << std::hex << decl.address << '_'
       << TranslateType(*type, dl, true);
 
-  // TODO(akshay): use `dl` to figure out the size in bytes of the type. Make
-  //               a `std::string` and initialize it to that size. Use the
-  //               `MemoryProvider` to query each byte offset from
-  //               `decl.address`, and if the byte is available, then write it
-  //               into the string. Inspect the permissions as you read/write
-  //               bytes. Be careful about crossing permission boundaries; i.e.
-  //               if we cross from writable into non-writable, then we should
-  //               treat that as a failure to get the needed bytes. Log a
-  //               warning/error, and we'll treat the var as an uninitialized
-  //               external. Use that same permissions tracking to set whether
-  //               or not the global var is constant. Try to factor this out
-  //               into one or more functions and not do it all right here.
-  //               Finally, if and when all bytes are read, and no permission
-  //               boundaries are crossed, go and invoke the data lifter.
-  const auto gv = new llvm::GlobalVariable(
-      *options.module, type, false, llvm::GlobalValue::ExternalLinkage,
-      llvm::Constant::getNullValue(type), ss2.str());
+  // Inspect the availability of first byte at `decl.address` and append
+  // it into the bytes vector
+  const auto data_size = dl.getTypeAllocSize(type);
+  auto [first_byte, first_byte_avail, first_byte_perms] =
+      memory_provider.Query(decl.address);
+  if (MemoryProvider::HasByte(first_byte_avail)) {
+    bytes.push_back(first_byte);
+    bytes_accessable = true;
+  }
 
-  ga->setAliasee(gv);
-  lifter_context.AddEntity(gv, decl.address);
+  // Inspect the read/write permission of bytes and check if it is crossing
+  // permission boundaries. Log error in such case
+  if (bytes_accessable) {
+    for (auto i = 1U; i < data_size; ++i) {
+      auto [byte, byte_avail, byte_perms] =
+          memory_provider.Query(decl.address + i);
+      if (!MemoryProvider::HasByte(byte_avail) ||
+          (first_byte_perms != byte_perms)) {
+
+        // Set to false if the byte is not available or it crosses
+        // permission boundary
+        bytes_accessable = false;
+        break;
+      }
+      bytes.push_back(byte);
+    }
+  }
 
-  return ga;
+
+  if (bytes_accessable) {
+    value = lifter_context.value_lifter.Lift(
+        std::string_view(reinterpret_cast<char *>(bytes.data()), bytes.size()),
+        type, lifter_context);
+  } else {
+    value = llvm::Constant::getNullValue(type);
```

Comment:
Here lets leave `value` as `nullptr`. Introduce a temporary "fixup" for this in `main` to initialize all globals.